### DISCUSSION TITLE
Is there a Linux driver for "soft" coreSPI?

### DISCUSSION THREAD

**Discussion by rary21** [polarfire-soc/polarfire-soc-documentation#472] [Linux drivers for PolarFire Soc] [answered]
Hi, I'd like to add a coreSPI module in fabric.

SPI module in MSS uses `spi-microchip-core.c` .
However, this driver seems incompatible with coreSPI found below
because INT_CLEAR register offset is different.
https://www.microsemi.com/document-portal/doc_view/1245299-corespi-hb
The commit message of `spi-microchip-core.c`  says that it is for "hard" controller 
so I guess there is no driver availble for coreSPI in fabric.

Is my understanding correct?

**Comment by eppidei** [comment, marked as answer]
Hi @rary21  you are correct. It will be available in next releases but I cannot give you a timeline at the moment.

**Reply by rary21** [reply]
Thank you! I got it.

**Reply by RossWilliamson** [reply]
Following as I also would like a driver for the soft coreSPI

**Reply by ddbabich** [reply]
Are there any updates on the release of a Linux soft core SPI driver?  I'm also interested.
Thanks

**Reply by RossWilliamson** [reply]
Is there any update when this driver will be availble?

**Comment by rockoguido** [comment]
hello for driver for devices in the fabric see 
https://github.com/Mi-V-Soft-RISC-V/miv-rv32-bare-metal-examples/tree/main/driver-examples/
amongst those you will find a corespi driver that uses the spi from the ESS.
These examples target a few of our boards, in particular our Eval board based on the 300T

**Reply by RossWilliamson** [reply]
Thanks - I think we are both looking for a linux kernel driver to expose it as a spidev device via device-tree

**Comment by rockoguido** [comment]
Hi...

I don't have any updates for you at this time...

Regards,


Dave Campanella

Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)

+1 760 304 0172 (office)
***@***.******@***.***>



[cid:3e56fdc9-4bc6-485d-b931-b095856fdffb]





________________________________

**Reply by RossWilliamson** [reply]
Just wondering if it was being actively worked on or if it would be something the community could help with (but not me I'm no good at Kernel development)

**Reply by rockoguido** [reply]
Hello.    Well I do have an example where I added a UART (16550) based into the fabric and can access from the MSS running Linux.   There are few steps...
1. Naturally on the fabric side had to add the extra logic to support this added UART, via the APB bus and using FIC3 for the MSS to Fabric Interconnect
2. On the Yocto Linux had to patch the device tree in order to add this UART , patch was mostly about the Interrupt Number and other UART related parameters
3. A new driver was not required since the other MSS Uarts are 16550 compliant.  
4. those are the steps in summary, my first thoughts about a SPI device in the fabric that can be access via Linux is that one would NOT use the ESS (which has SPI capability along with other serialized devices), you would pull in a dedicated SPI based IP that hopefully is based on the SPI block within the MSS, that way the driver is mostly done,   I have yet to look at this but it is on our list.
5. I hope that gives you a little insight
6. Regards.
7. Dave Campanella

**Reply by Quinn17** [reply]
I Dave,

I was wondering if you could share the device tree patch you used for the UART. I am trying to do the same thing (using FIC3 CoreUARTTapbC0) but I am struggling. I'm using an overlay to test it first, but it doesn't seem to work (I see it binding on ttyS2 but the message I send are received as `/0/0/0` from my analyzer). Here is my overlay : 
```
/dts-v1/;
/plugin/;

/ {
    compatible = "microchip,mpfs";

    fragment@0 {
        target-path = "/fabric-bus@40000000";
        __overlay__ {
            uart@40000300 {
                compatible = "ns16550a";
                reg = <0x00 0x40000300 0x00 0x100>;
                reg-shift = <2>;                  // Common for memory-mapped 16550
                reg-io-width = <4>;               // 4 bytes per register
                clock-frequency = <50000000>;     // Adjust based on your IP configuration
                current-speed = <115200>;
                status = "okay";
            };
        };
    };
};
```

Thanks,
Quentin

PS : I am new to Yocto and Linux device trees.

**Comment by prajnark** [comment]
Hi @rary21  @RossWilliamson,

Support for the **Soft CoreSPI** Linux driver is now available in Microchip’s linux4microchip kernel tree, and on the master branches of Buildroot and Yocto.

Linux kernel tree: https://github.com/linux4microchip/linux
Buildroot: https://github.com/linux4microchip/buildroot-external-microchip
Yocto: https://github.com/linux4microchip/meta-mchp

Best regards,
Prajna

**Reply by RossWilliamson** [reply]
Thanks! I'm having some trouble getting the driver to initialize.  Is this the correct device tree entry to get the softcore to work? 

```
 spi@40000400 {
                        #address-cells = <0x01>;
                        compatable = "microchip,corespi-rtl-v5";
                        clocks = <0x16 0x07>;
                        #size-cells = <0x00>;
                        interrupt-parent = <0x09>;
                        status = "okay";
                        reg = <0x00 0x40000400 0x00 0x1000>;

                        spidev@0 {
                                spi-max-frequency = <0x7a120>;
                                compatible = "linux,spidev\0rohm,bh2228fv";
                                status = "okay";
                                reg = <0x00>;
                        };
                };
```

**Reply by RossWilliamson** [reply]
Nevermind - I can't spell compatible. I do have an issue with a kernel panic but I've posted that [here](https://github.com/polarfire-soc/platform/issues/31)
